mm2src/common/custom_futures/repeatable.rs

```diff
@@ -0,0 +1,785 @@
+        /// An error occurred during the last attempt.
+        error: E,
+    },
+}
+
+impl<E> RepeatError<E> {
+    pub fn error(&self) -> &E {
+        match self {
+            RepeatError::TimeoutExpired { error, .. } | RepeatError::AttemptsExceed { error, .. } => error,
+        }
+    }
+
+    pub fn into_error(self) -> E {
+        match self {
+            RepeatError::TimeoutExpired { error, .. } | RepeatError::AttemptsExceed { error, .. } => error,
+        }
+    }
+
+    fn timeout(until: Until, error: E) -> Self { RepeatError::TimeoutExpired { until, error } }
+
+    fn attempts(attempts: usize, error: E) -> Self { RepeatError::AttemptsExceed { attempts, error } }
+}
+
+impl<E: fmt::Display> fmt::Display for RepeatError<E> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            RepeatError::TimeoutExpired { until, error } => {
+                write!(
+                    f,
+                    "Waited too long until {until:?} for the future to succeed. Error: {error}",
+                )
+            },
+            RepeatError::AttemptsExceed { attempts, error } => {
+                write!(f, "Error {error} on retrying the future after {attempts} attempts")
+            },
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum Until {
+    Instant(Instant),
+    TimestampMs(u64),
+}
+
+/// The future is ether ready (with a `T` result), or not ready (failed with an intermediate `E` error).
+#[derive(Debug)]
+pub enum Action<T, E> {
+    Ready(T),
+    Retry(E),
+}
+
+pub trait RetryOnError<T, E> {
+    fn retry_on_err(self) -> Action<T, E>;
+}
+
+impl<T, E> RetryOnError<T, E> for Result<T, E> {
+    /// Converts `Result<T, E>` into `Action<T, E>`:
+    /// * `Ok(T)` => `Action::Ready(T)`.
+    /// * `Err(E)` => `Action::Retry(E)`.
+    #[inline]
+    fn retry_on_err(self) -> Action<T, E> {
+        match self {
+            Ok(ready) => Action::Ready(ready),
+            Err(e) => Action::Retry(e),
+        }
+    }
+}
+
+/// The result of `repeatable` or `retry_on_err` macros - the first step at the future configuration.
+pub struct Repeatable<Factory, F, T, E> {
+    factory: Factory,
+    /// Currently executable future, i.e. an active attempt.
+    exec_fut: Option<F>,
+    /// A timeout future if we're currently waiting for a timeout.
+    timeout_fut: Option<Timer>,
+    until: Option<Until>,
+    attempts: Option<AttemptsState>,
+    repeat_every: Option<Duration>,
+    inspect_err: Option<Box<dyn InspectErrorTrait<E>>>,
+    _phantom: PhantomData<(F, T, E)>,
+}
+
+impl<Factory, F, T, E> Repeatable<Factory, F, T, E>
+where
+    Factory: FactoryTrait<F>,
+    F: RepeatableTrait<T, E>,
+{
+    #[inline]
+    pub fn new(factory: Factory) -> Self {
+        let exec = factory();
+
+        Repeatable {
+            factory,
+            exec_fut: Some(exec),
+            timeout_fut: None,
+            until: None,
+            attempts: None,
```

Comment:
As default behavior is to always perform at least one attempt, it's worth making it non-Option and initializing with
```rust
AttemptsState {
    current_attempt: 0,
    total_attempts: 1,
}
```

Reply:
I agree, will do

Reply:
I've checked, and we can't avoid using `Option<AttemptsState>`, because if the `until` is specified only, then there will be only one attempt due to the default `AttemptsState::total_attempts = 1`:
```
Actual:  Err(AttemptsExceed { attempts: 1, error: "Not ready" })
Expected: Err(TimeoutExpired { until_ms: 1671809910391, error: "Not ready" })
```

Reply:
Are we going to use attempts limit and `until_ms` *together*? If no, then it might be worth using the following:
```rust
enum RepeatUntil {
    TimeoutReached(u64),
    AttemptsReached(usize)
}
```

This way, we will clearly define that `Repeatable` should either have a timeout or attempts limit.

Reply:
We discussed this at the thread https://github.com/KomodoPlatform/atomicDEX-API/pull/1564#discussion_r1040959467
Anyway this makes sense, so I can refactor the behaviour

Reply:
I see, the decision was to perform at least one attempt always, and it can be explicitly set using `RepeatUntil::AttemptsReached(1)` as default. Hopefully, it will be a bit more clear this way 🙂

Reply:
Done